src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -672,7 +687,10 @@ private static float CosineSimilarityCore(ReadOnlySpan<float> x, ReadOnlySpan<fl
                     if (i != x.Length)
                     {
                         uint lastVectorIndex = (uint)(x.Length - Vector512<float>.Count);
-                        TUnaryOperator.Invoke(Vector512.LoadUnsafe(ref xRef, lastVectorIndex)).StoreUnsafe(ref dRef, lastVectorIndex);
+                        Vector512.ConditionalSelect(
+                            Vector512.Equals(LoadRemainderMaskSingleVector512(x.Length - i), Vector512<float>.Zero),
+                            Vector512.LoadUnsafe(ref dRef, lastVectorIndex),
```

Comment:
Is it worth avoiding this load by caching the last vector computed from the main loop?

Reply:
I'm not following; where did we already load this vector?

Reply:
We're basically backtracking by `Count - remainder`, doing an operation, and then blending the already computed entries with the new entries before storing.

Say we have a total number `19` inputs. We'd do 4 full loop iterations (`0-3`, `4-7`, `8-11`, `12-15`).

We'd then have `3` remaining, so we backtrack by 1 and reprocess `15-18`. To support exact overlapping input/output, we need to preserve the already computed `15` which already exists as the last element from the previous loop iteration.

So, rather than doing another memory access, we could just cache the last stored vector from the loop, shuffle the elements left by 3, and use that in the conditional select.

Reply:
That might be something worth doing here, but it might also be fine to postpone that for "modern .NET only" when we can get some helper API in place to make that process simpler.

Reply:
Ok, let's leave it as is for now.